src/xdp-sealed-fd.c

```diff
@@ -0,0 +1,226 @@
+#include <gio/gunixfdlist.h>
+#include <sys/mman.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "xdp-utils.h"
+#include "xdp-sealed-fd.h"
+
+static const int required_seals = F_SEAL_GROW | F_SEAL_WRITE | F_SEAL_SHRINK;
+
+struct _XdpSealedFd {
+  GObject parent_instance;
+
+  int fd;
+  gboolean is_sealed;
+};
+
+G_DEFINE_FINAL_TYPE (XdpSealedFd, xdp_sealed_fd, G_TYPE_OBJECT)
+
+static void
+xdp_sealed_fd_finalize (GObject *object)
+{
+  XdpSealedFd *sealed_fd = XDP_SEALED_FD (object);
+
+  xdp_close_fd (&sealed_fd->fd);
+
+  G_OBJECT_CLASS (xdp_sealed_fd_parent_class)->finalize (object);
+}
+
+static void
+xdp_sealed_fd_class_init (XdpSealedFdClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->finalize = xdp_sealed_fd_finalize;
+}
+
+static void
+xdp_sealed_fd_init (XdpSealedFd *sealed_fd)
+{
+  sealed_fd->fd = -1;
+}
+
+XdpSealedFd *
+xdp_sealed_fd_new_take_memfd (int      memfd,
+                              GError **error)
+{
+  int saved_errno = -1;
+  g_autoptr(XdpSealedFd) sealed_fd = NULL;
+  xdp_autofd int fd = memfd;
+  int seals;
+
+  g_return_val_if_fail (fd != -1, NULL);
+
+  if ((seals = fcntl (fd, F_GET_SEALS)) == -1)
+   {
+      saved_errno = errno;
+
+      g_set_error (error,
+                   G_IO_ERROR,
+                   g_io_error_from_errno (saved_errno),
+                   "fcntl F_GET_SEALS: %s", g_strerror (saved_errno));
+      return NULL;
+    }
+
+  /* If the seal seal is set and some required seal is missing report EPERM error directly */
+  if ((seals & F_SEAL_SEAL) && (seals & required_seals) != required_seals)
+    saved_errno = EPERM;
+  else if (fcntl (fd, F_ADD_SEALS, required_seals) == -1)
+    saved_errno = errno;
+
+  if (saved_errno != -1)
+    {
+      g_set_error (error,
+                   G_IO_ERROR,
+                   g_io_error_from_errno (saved_errno),
+                   "fcntl F_ADD_SEALS: %s", g_strerror (saved_errno));
+      return NULL;
+    }
+
+  sealed_fd = g_object_new (XDP_TYPE_SEALED_FD, NULL);
+  sealed_fd->fd = xdp_steal_fd (&fd);
+
+  return g_steal_pointer (&sealed_fd);
+}
+
+XdpSealedFd *
+xdp_sealed_fd_new_from_bytes (GBytes  *bytes,
+                              GError **error)
+{
+  gconstpointer bytes_data;
+  gsize bytes_len;
+  xdp_autofd int fd;
+  int saved_errno = -1;
+  g_autoptr(XdpSealedFd) sealed_fd = NULL;
+  char *shm;
+  g_autoptr(GOutputStream) stream = NULL;
+
+  g_return_val_if_fail (bytes != NULL, NULL);
+
+  if ((fd = memfd_create ("sealed-fd-from-bytes", MFD_ALLOW_SEALING)) == -1)
+    {
+      int saved_errno;
+
+      saved_errno = errno;
+      g_set_error (error,
+                   G_IO_ERROR,
+                   g_io_error_from_errno (saved_errno),
+                   "memfd_create: %s", g_strerror (saved_errno));
+      return NULL;
+    }
+
+  bytes_data = g_bytes_get_data (bytes, &bytes_len);
+
+  if ((shm = (char *) mmap (NULL, bytes_len, PROT_WRITE, MAP_SHARED, fd, 0)) == (void *) -1)
```

Comment:
Why is this `MAP_SHARED`?

Reply:
we need it so that changes are written back to the file. We can't use `MAP_PRIVATE` because it is copy-on-write. See https://www.man7.org/linux/man-pages/man2/mmap.2.html

Reply:
Ah yes, we use this  to write to the memfd so GMappedFile also won't work here...

Reply:
i find it odd that we have the write flag but it doesn't mean that data is written back to the file.